### PR TITLE
(sentry) update `mark-as-read` api to be called from saga instead of `componentDidUpdate()`

### PR DESCRIPTION
**Pull request by ratik21**

### CODE REVIEW

**Review by domw30 — APPROVED** [review]
Looks good to me. Lets get @dalefukami's approval too

File: src/store/messages/saga.ts
Comment (dalefukami) [inline src/store/messages/saga.ts:118]: Let's add a type here. I know we still have a few of these events that don't specify a type but we should try to make the new ones use it. It allows us to filter properly in the listeners (rather than listening for '*')

File: src/store/channels/saga.ts
Comment (dalefukami) [inline src/store/channels/saga.ts:69]: I'm trying to decide if this is the right event. If we start pre-loading messages for some reason, for example, then we'd be marking the messages as read all incorrectly for some things. Maybe we just need to make this more strict and only react if the channel matches the open conversation / channel?

**Review by dalefukami — COMMENTED** [review]
:heart: the move to sagas!!

File: src/store/messages/saga.ts
Comment (ratik21) [reply src/store/messages/saga.ts:118]: makes sense, updated. Will keep in mind for the new ones.

File: src/store/channels/saga.ts
Comment (dalefukami) [reply src/store/channels/saga.ts]: Also just had another thought, we also need to mark as read when new real time messages come in and the conversation is open.

Comment (ratik21) [reply src/store/channels/saga.ts]: I know about the `activeConversationId`, but i am not sure if we have an "active channel id"..? Or can we just extract/know that from the url maybe? eg

image

Comment (dalefukami) [reply src/store/channels/saga.ts]: Yeah, I'm not sure, we'd have to dig around.  But somehow we must know because we load the messages for the channel and receive real time messages, etc. Maybe it mostly currently works just based on the fact that it's open? We should probably have that in state somewhere.

Comment (ratik21) [reply src/store/channels/saga.ts]: So this is getting the channelId from the current url i think, and passing it via props in the UI

https://github.com/zer0-os/zOS/blob/a51af1a8074d6efe3cdc68ad5d657066e5282a55/src/platform-apps/channels/index.tsx#L19

> But somehow we must know because we load the messages for the channel and receive real time messages, etc

correct.

Comment (ratik21) [reply src/store/channels/saga.ts]: I will update to read the "current channel id" from the url. And also trigger `mark-as-read` on receiveNewMessage sendbird event

Comment (dalefukami) [reply src/store/channels/saga.ts]: In this case I think maybe we need to have the component tell us when the channel changes. All the routing is done via React so it makes sense that it is the original source of that state. I.E., when the channel changes in the main component (wherever that is) it can call a "setCurrentChannel" or something that will just update the global state.

Comment (ratik21) [reply src/store/channels/saga.ts]: I have updated the PR as per your comments. But the code got a little more complex that i would like 🤔

File: src/components/chat-view-container/chat-view-container.test.tsx
Comment (dalefukami) [inline src/components/chat-view-container/chat-view-container.test.tsx:25]: can we remove this now?

File: src/store/messages/saga.ts
Comment (dalefukami) [inline src/store/messages/saga.ts:370]: I don't think this is right. We only want to mark the messages read if it's the conversation/channel that's currently open.

File: src/components/chat-view-container/chat-view-container.tsx
Comment (dalefukami) [inline src/components/chat-view-container/chat-view-container.tsx:101]: Wouldn't this be set if you had both channels and the chat window open? So, if you have the chat window open we wouldn't change the activeChannelId correctly?

Also, do we clear the activeChannelId somewhere when we navigate away from the channels route?

I think maybe managing the activeChannelId could be it's own PR so we can figure out the details there first.

File: src/store/channels/saga.ts
Comment (dalefukami) [inline src/store/channels/saga.ts:30]: Isn't the user id supposed to be the currently logged in user? We're saying, "this user has read these messages". So, admin messages shouldn't have any impact here.

We probably still need this in the case of public viewers because they won't have a userId but I'm worried we're passing the wrong user id here maybe?

File: src/store/channels/saga.ts
Comment (dalefukami) [inline src/store/channels/saga.ts:98]: I think maybe we should have 2 events. One for Channels and one for Conversations. I know they got merged together but we should treat them as totally different things.

Also, we shouldn't mark everything read like this. We should only mark the conversations as read IF this event was for the already open channel. The bug here is, what if we decided to start preloading messages for channels that we're not viewing at the moment? We'd get a bunch of these events but we wouldn't want to just "mark the current conversation" unread.

File: src/store/messages/saga.ts
Comment (dalefukami) [reply src/store/messages/saga.ts:370]: Ah, this is also where we're using the wrong userId. Maybe just don't send the userId and let the other saga fetch the current user.

File: src/components/chat-view-container/chat-view-container.tsx
Comment (dalefukami) [reply src/components/chat-view-container/chat-view-container.tsx:101]: Actually, maybe this should be one or two steps higher where we're specifically rendering the Channel. Eg: the `platform-app/channels/container.tsx` component.

File: src/store/channels/saga.ts
Comment (dalefukami) [inline src/store/channels/saga.ts:96]: Wonder if we can use the `takeEveryFromBus` helper that I created a couple weeks ago?

**Review by dalefukami — CHANGES_REQUESTED** [review]

File: src/store/channels/saga.ts
Comment (ratik21) [reply src/store/channels/saga.ts:96]: Works! Thankss

File: src/components/chat-view-container/chat-view-container.tsx
Comment (ratik21) [reply src/components/chat-view-container/chat-view-container.tsx:101]: created a separate PR for it

**Review by dalefukami — APPROVED** [review]
Looks good once the other PR is merged into `main`. Nicely done.

File: src/store/messages/saga.ts
Comment (ratik21) [reply src/store/messages/saga.ts:384]: I thought this would be..automatically handled..? Since we're marking as read only for active channel/conversation.

File: src/store/channels/saga.ts
Comment (ratik21) [reply src/store/channels/saga.ts:28]: will add a check. but this makes me realize i didn't test on the public zos mode. How do we mark those messages as read (currently) if the user-id isn't defined .. 🤔

File: src/store/messages/saga.ts
Comment (ratik21) [reply src/store/messages/saga.ts:384]: also if we pass the channelId, wouldn't i mark the messages as read for a channel which isn't "active" (i.e receiveNewMessage on *channelX* but i am on *channelY*)

Comment (dalefukami) [reply src/store/messages/saga.ts]: But those sagas both have that check already, don't the?

File: src/store/channels/saga.ts
Comment (dalefukami) [reply src/store/channels/saga.ts:28]: We don't. There's no unread messages in public zos

File: src/store/channels/saga.ts
Comment (dalefukami) [inline src/store/channels/saga.ts:59]: Ah, I just realized these aren't as specific as I thought. This function should take the channel id and ONLY call `markAllMessagesAsRead` if the id passed in matches the active conversation. (and likewise for the channel one). Something like, `markUnreadIfActive(channelId)`

Comment (dalefukami) [reply src/store/channels/saga.ts:59]: Otherwise, we'll be marking the active channel as read every time anything loads in the background.

File: src/store/messages/saga.ts
Comment (ratik21) [reply src/store/messages/saga.ts]: no they don't 👀 

bdw i am not sure how to use `takeEveryFromBus` in case i wanted the event to return a data, and pass that data to the calling function (in this case the `channelId`)
image

Comment (dalefukami) [reply src/store/messages/saga.ts]: Something like:
```
yield takeEveryFromBus(
  yield call(conversationsChannel),
  ChannelEvents.MessagesLoadedForConversation,
  (action) => markAllMessagesAsReadIfActive(action.channelId),
);
```

File: src/store/channels/saga.ts
Comment (ratik21) [reply src/store/channels/saga.ts:59]: :+1:, updated

**Review by dalefukami — APPROVED** [review]

**Pull request merged**
